### DISCUSSION TITLE
Concatenate files to build a script, then use script as a test

### DISCUSSION THREAD

**Discussion by andy5995** [mesonbuild/meson#11573] [Q&A]
I'm trying to concatenate two files to make a full script, but I don't know why this isn't working:

```
newscripts = [
  'fail_restore_files_not_in_waste',
  '01_arg_u',
  'mrl_not_exist',
  'restore_with_wildcard'
  ]

build_test_root = meson.current_build_dir()
RMW_FAKE_HOME = join_paths(build_test_root, 'rmw-tests-home')

cat_prog = find_program('cat')

foreach scriptfoo : newscripts
  concat_fragments = custom_target(
    'concat_fragments',
    command: [ cat_prog, '@INPUT@' ],
    capture: true,
    input: files(
      join_paths(meson.current_source_dir(), 'script-head'),
      join_paths(meson.current_source_dir(), 'fragments', scriptfoo + '.sh')),
    output: scriptfoo + '.sh',
    install: false,
    build_by_default: true
  )

  test(
    scriptfoo,
    files(join_paths(build_test_root, scriptfoo + '.sh')),
    env : {
      'MESON_SOURCE_ROOT': meson.project_source_root(),
      'RMW_FAKE_HOME': RMW_FAKE_HOME,
      'MESON_BUILD_ROOT': meson.project_build_root()
      },
    depends: [main_bin, concat_fragments]
  )
endforeach
```

The error is 

../test/meson.build:38:2: ERROR: File /home/andy/src/rmw-project/rmw/builddir/test/fail_restore_files_not_in_waste.sh does not exist.

I have the target set in the `depends` field, so I don't know why it's not getting built so the test can run.

**Comment by eli-schwartz** [comment]
Because files are not built before the meson.build file is parsed. I am not actually sure why you're using files() at all here.

The custom_target "input" kwarg expects a list of things that coerce to files, which includes just using the strings:

```meson
input: ['script-head', 'fragments/@0@.sh'.format(scriptfoo)]
```

Note that the `files()` function is only permitted to be used for files that already exist *before* meson is run, so you are not permitted to use it on files in the builddir.

If you're trying to work around the fact that `test()` doesn't type-check and permit a custom target output as the command, then the bigger problem might be the fact that it isn't executable... So you'd need to pass it as an argument to e.g. `find_program('sh')`, which does work -- stick sh as the test program and the script as the args.

**Reply by andy5995** [reply]
I want to concatenate 'script-head' with each script "fragment" in the source tests/fragments directory, and then output each of the concatenated file to the test/ directory under the builddir.

What I have so far: https://github.com/theimpossibleastronaut/rmw/pull/397
